File: test_api.js
/**
 * CheqPrint API - Script de test Node.js
 * =======================================
 *
 * Usage:
 *    1. Créez un fichier .env avec votre clé API:
 *       API_KEY=sk_live_VOTRE_CLE
 *
 *    2. Installez les dépendances:
 *       npm install
 *
 *    3. Lancez le script:
 *       node test_api.js
 */

require('dotenv').config();
const readline = require('readline');

// Configuration
const BASE_URL = "https://xwpgblfdmlrrkksmuazy.supabase.co/functions/v1";
const API_KEY = process.env.API_KEY || "";

// Couleurs
const colors = {
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  reset: '\x1b[0m',
  bold: '\x1b[1m'
};

function printHeader(text) {
  console.log(`\n${colors.bold}${colors.blue}${'='.repeat(60)}${colors.reset}`);
  console.log(`${colors.bold}${colors.blue}  ${text}${colors.reset}`);
  console.log(`${colors.bold}${colors.blue}${'='.repeat(60)}${colors.reset}\n`);
}

function printSuccess(text) {
  console.log(`${colors.green}✓ ${text}${colors.reset}`);
}

function printError(text) {
  console.log(`${colors.red}✗ ${text}${colors.reset}`);
}

function printInfo(text) {
  console.log(`${colors.yellow}→ ${text}${colors.reset}`);
}

async function makeRequest(method, endpoint, data = null, auth = true) {
  const url = `${BASE_URL}/${endpoint}`;
  const headers = { "Content-Type": "application/json" };

  if (auth && API_KEY) {
    headers["Authorization"] = `Bearer ${API_KEY}`;
  }

  try {
    const options = {
      method,
      headers
    };

    if (data) {
      options.body = JSON.stringify(data);
    }

    const response = await fetch(url, options);
    const responseData = await response.json().catch(() => ({}));

    return {
      status: response.status,
      data: responseData,
      success: response.status < 400
    };
  } catch (error) {
    return {
      status: 0,
      data: { error: error.message },
      success: false
    };
  }
}

async function testBanks() {
  printInfo("GET /banks - Liste des banques disponibles");

  const result = await makeRequest("GET", "banks", null, false);

  if (result.success) {
    const banks = result.data.data || [];
    printSuccess(`Trouvé ${banks.length} banque(s)`);
    banks.slice(0, 5).forEach(bank => {
      console.log(`   - ${bank.name || bank.code || 'N/A'}`);
    });
  } else {
    printError(`Erreur: ${JSON.stringify(result.data)}`);
  }

  return result.success;
}

async function testTemplates() {
  printInfo("GET /templates - Templates de chèques");

  const result = await makeRequest("GET", "templates");

  if (result.success) {
    const templates = result.data.data || [];
    printSuccess(`Trouvé ${templates.length} template(s)`);
  } else {
    printError(`Erreur: ${JSON.stringify(result.data)}`);
  }

  return result.success;
}

async function testCompanies() {
  printInfo("GET /companies - Vos sociétés");

  const result = await makeRequest("GET", "companies");

  if (result.success) {
    const companies = result.data.data || [];
    printSuccess(`Trouvé ${companies.length} société(s)`);
    companies.slice(0, 5).forEach(company => {
      console.log(`   - ${company.name || 'N/A'}`);
    });
    return companies;
  } else {
    printError(`Erreur: ${JSON.stringify(result.data)}`);
    return [];
  }
}

async function testChequeBooks() {
  printInfo("GET /cheque-books - Vos carnets de chèques");

  const result = await makeRequest("GET", "cheque-books");

  if (result.success) {
    const books = result.data.data || [];
    printSuccess(`Trouvé ${books.length} carnet(s)`);
    books.slice(0, 5).forEach(book => {
      const remaining = (book.totalCheques || 0) - (book.usedCheques || 0);
      const status = book.isActive ? "actif" : "inactif";
      console.log(`   - ${book.companyName} / ${book.bankName} (${remaining} restants, ${status})`);
    });
    return books;
  } else {
    printError(`Erreur: ${JSON.stringify(result.data)}`);
    return [];
  }
}

async function testPrintHistory() {
  printInfo("GET /print-history - Historique des impressions");

  const result = await makeRequest("GET", "print-history?limit=5");

  if (result.success) {
    const data = result.data.data || {};
    const records = data.records || [];
    const total = data.total || 0;
    printSuccess(`Total: ${total} impression(s)`);
    records.slice(0, 5).forEach(record => {
      console.log(`   - Chèque #${record.chequeNumber} : ${record.beneficiary} - ${record.amount} FCFA`);
    });
  } else {
    printError(`Erreur: ${JSON.stringify(result.data)}`);
  }

  return result.success;
}

async function testPrintCheque(companyName, bankName) {
  printInfo("POST /print-cheque - Imprimer un chèque");

  const data = {
    beneficiaire: "TEST API NODEJS",
    montant: "234567",
    selectedBank: bankName,
    companyName: companyName,
    lieu: "ABIDJAN"
  };

  console.log(`   Données: ${JSON.stringify(data, null, 2)}`);

  const result = await makeRequest("POST", "print-cheque", data);

  if (result.success) {
    const cheque = result.data.data || {};
    printSuccess("Chèque imprimé avec succès!");
    console.log(`   Numéro: ${cheque.chequeNumber}`);
    console.log(`   Montant en lettres: ${cheque.montantEnLettres}`);
  } else {
    printError(`Erreur: ${JSON.stringify(result.data)}`);
  }

  return result.success;
}

async function testPrintBatch(companyName, bankName) {
  printInfo("POST /print-batch - Imprimer plusieurs chèques");

  const data = {
    selectedBank: bankName,
    companyName: companyName,
    cheques: [
      { beneficiaire: "BATCH JS 1", montant: "60000" },
      { beneficiaire: "BATCH JS 2", montant: "85000" }
    ]
  };

  console.log(`   Nombre de chèques: ${data.cheques.length}`);

  const result = await makeRequest("POST", "print-batch", data);

  if (result.success) {
    const batch = result.data.data || {};
    printSuccess(`Lot imprimé: ${batch.totalProcessed} chèques`);
    console.log(`   Montant total: ${batch.totalAmount} FCFA`);
    (batch.cheques || []).forEach(cheque => {
      console.log(`   - #${cheque.chequeNumber}: ${cheque.beneficiaire} - ${cheque.montant} FCFA`);
    });
  } else {
    printError(`Erreur: ${JSON.stringify(result.data)}`);
  }

  return result.success;
}

function askQuestion(question) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  return new Promise(resolve => {
    rl.question(question, answer => {
      rl.close();
      resolve(answer);
    });
  });
}

async function main() {
  printHeader("CheqPrint API - Tests Node.js");

  // Vérifier la clé API
  if (!API_KEY) {
    printError("Clé API non configurée!");
    printInfo("Créez un fichier .env avec: API_KEY=sk_live_VOTRE_CLE");
    printInfo("Ou définissez la variable d'environnement API_KEY");
    return;
  }

  printInfo(`Clé API: ${API_KEY.substring(0, 20)}...`);
  printInfo(`URL: ${BASE_URL}`);

  // Tests de lecture
  printHeader("Tests de lecture");

  await testBanks();
  await testTemplates();
  await testCompanies();
  const books = await testChequeBooks();
  await testPrintHistory();

  // Trouver un carnet actif
  let activeBook = null;
  for (const book of books) {
    if (book.isActive) {
      const remaining = (book.totalCheques || 0) - (book.usedCheques || 0);
      if (remaining >= 3) {
        activeBook = book;
        break;
      }
    }
  }

  if (activeBook) {
    printHeader("Tests d'écriture");
    printInfo(`Utilisation du carnet: ${activeBook.companyName} / ${activeBook.bankName}`);

    const response = await askQuestion("\n⚠️  Ces tests vont consommer des numéros de chèques. Continuer? (o/N): ");

    if (response.toLowerCase() === 'o') {
      await testPrintCheque(activeBook.companyName, activeBook.bankName);
      await testPrintBatch(activeBook.companyName, activeBook.bankName);
    } else {
      printInfo("Tests d'écriture ignorés");
    }
  } else {
    printHeader("Tests d'écriture");
    printError("Aucun carnet actif avec assez de chèques trouvé");
    printInfo("Créez un carnet de chèques dans l'app CheqPrint");
  }

  printHeader("Tests terminés");
}

main().catch(console.error);

File: test_api.py
#!/usr/bin/env python3
"""
CheqPrint API - Script de test Python
======================================

Usage:
    1. Créez un fichier .env avec votre clé API:
       API_KEY=sk_live_VOTRE_CLE

    2. Installez les dépendances:
       pip install requests python-dotenv

    3. Lancez le script:
       python test_api.py
"""

import os
import json
import requests
from datetime import datetime

# Essayer de charger python-dotenv si disponible
try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass

# Configuration
BASE_URL = "https://xwpgblfdmlrrkksmuazy.supabase.co/functions/v1"
API_KEY = os.getenv("API_KEY", "")

# Couleurs pour le terminal
class Colors:
    GREEN = '\033[92m'
    RED = '\033[91m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    RESET = '\033[0m'
    BOLD = '\033[1m'

def print_header(text):
    print(f"\n{Colors.BOLD}{Colors.BLUE}{'='*60}{Colors.RESET}")
    print(f"{Colors.BOLD}{Colors.BLUE}  {text}{Colors.RESET}")
    print(f"{Colors.BOLD}{Colors.BLUE}{'='*60}{Colors.RESET}\n")

def print_success(text):
    print(f"{Colors.GREEN}✓ {text}{Colors.RESET}")

def print_error(text):
    print(f"{Colors.RED}✗ {text}{Colors.RESET}")

def print_info(text):
    print(f"{Colors.YELLOW}→ {text}{Colors.RESET}")

def make_request(method, endpoint, data=None, auth=True):
    """Effectue une requête à l'API"""
    url = f"{BASE_URL}/{endpoint}"
    headers = {"Content-Type": "application/json"}

    if auth and API_KEY:
        headers["Authorization"] = f"Bearer {API_KEY}"

    try:
        if method == "GET":
            response = requests.get(url, headers=headers)
        elif method == "POST":
            response = requests.post(url, headers=headers, json=data)
        else:
            raise ValueError(f"Méthode non supportée: {method}")

        return {
            "status": response.status_code,
            "data": response.json() if response.text else {},
            "success": response.status_code < 400
        }
    except requests.exceptions.RequestException as e:
        return {
            "status": 0,
            "data": {"error": str(e)},
            "success": False
        }

def test_banks():
    """Test 1: Liste des banques (sans authentification)"""
    print_info("GET /banks - Liste des banques disponibles")

    result = make_request("GET", "banks", auth=False)

    if result["success"]:
        banks = result["data"].get("data", [])
        print_success(f"Trouvé {len(banks)} banque(s)")
        for bank in banks[:5]:  # Afficher max 5
            print(f"   - {bank.get('name', bank.get('code', 'N/A'))}")
    else:
        print_error(f"Erreur: {result['data']}")

    return result["success"]

def test_templates():
    """Test 2: Liste des templates"""
    print_info("GET /templates - Templates de chèques")

    result = make_request("GET", "templates")

    if result["success"]:
        templates = result["data"].get("data", [])
        print_success(f"Trouvé {len(templates)} template(s)")
    else:
        print_error(f"Erreur: {result['data']}")

    return result["success"]

def test_companies():
    """Test 3: Liste des sociétés"""
    print_info("GET /companies - Vos sociétés")

    result = make_request("GET", "companies")

    if result["success"]:
        companies = result["data"].get("data", [])
        print_success(f"Trouvé {len(companies)} société(s)")
        for company in companies[:5]:
            print(f"   - {company.get('name', 'N/A')}")
        return companies
    else:
        print_error(f"Erreur: {result['data']}")
        return []

def test_cheque_books():
    """Test 4: Liste des carnets de chèques"""
    print_info("GET /cheque-books - Vos carnets de chèques")

    result = make_request("GET", "cheque-books")

    if result["success"]:
        books = result["data"].get("data", [])
        print_success(f"Trouvé {len(books)} carnet(s)")
        for book in books[:5]:
            remaining = book.get('totalCheques', 0) - book.get('usedCheques', 0)
            status = "actif" if book.get('isActive') else "inactif"
            print(f"   - {book.get('companyName')} / {book.get('bankName')} ({remaining} restants, {status})")
        return books
    else:
        print_error(f"Erreur: {result['data']}")
        return []

def test_print_history():
    """Test 5: Historique des impressions"""
    print_info("GET /print-history - Historique des impressions")

    result = make_request("GET", "print-history?limit=5")

    if result["success"]:
        data = result["data"].get("data", {})
        records = data.get("records", [])
        total = data.get("total", 0)
        print_success(f"Total: {total} impression(s)")
        for record in records[:5]:
            print(f"   - Chèque #{record.get('chequeNumber')} : {record.get('beneficiary')} - {record.get('amount')} FCFA")
    else:
        print_error(f"Erreur: {result['data']}")

    return result["success"]

def test_print_cheque(company_name, bank_name):
    """Test 6: Imprimer un chèque"""
    print_info("POST /print-cheque - Imprimer un chèque")

    data = {
        "beneficiaire": "TEST API PYTHON",
        "montant": "123456",
        "selectedBank": bank_name,
        "companyName": company_name,
        "lieu": "ABIDJAN"
    }

    print(f"   Données: {json.dumps(data, indent=2)}")

    result = make_request("POST", "print-cheque", data)

    if result["success"]:
        cheque = result["data"].get("data", {})
        print_success("Chèque imprimé avec succès!")
        print(f"   Numéro: {cheque.get('chequeNumber')}")
        print(f"   Montant en lettres: {cheque.get('montantEnLettres')}")
    else:
        print_error(f"Erreur: {result['data']}")

    return result["success"]

def test_print_batch(company_name, bank_name):
    """Test 7: Imprimer plusieurs chèques"""
    print_info("POST /print-batch - Imprimer plusieurs chèques")

    data = {
        "selectedBank": bank_name,
        "companyName": company_name,
        "cheques": [
            {"beneficiaire": "BATCH TEST 1", "montant": "50000"},
            {"beneficiaire": "BATCH TEST 2", "montant": "75000"}
        ]
    }

    print(f"   Nombre de chèques: {len(data['cheques'])}")

    result = make_request("POST", "print-batch", data)

    if result["success"]:
        batch = result["data"].get("data", {})
        print_success(f"Lot imprimé: {batch.get('totalProcessed')} chèques")
        print(f"   Montant total: {batch.get('totalAmount')} FCFA")
        for cheque in batch.get("cheques", []):
            print(f"   - #{cheque.get('chequeNumber')}: {cheque.get('beneficiaire')} - {cheque.get('montant')} FCFA")
    else:
        print_error(f"Erreur: {result['data']}")

    return result["success"]

def main():
    print_header("CheqPrint API - Tests Python")

    # Vérifier la clé API
    if not API_KEY:
        print_error("Clé API non configurée!")
        print_info("Créez un fichier .env avec: API_KEY=sk_live_VOTRE_CLE")
        print_info("Ou définissez la variable d'environnement API_KEY")
        return

    print_info(f"Clé API: {API_KEY[:20]}...")
    print_info(f"URL: {BASE_URL}")

    # Tests de lecture (sans modification)
    print_header("Tests de lecture")

    test_banks()
    test_templates()
    companies = test_companies()
    books = test_cheque_books()
    test_print_history()

    # Trouver un carnet actif pour les tests d'écriture
    active_book = None
    for book in books:
        if book.get("isActive"):
            remaining = book.get("totalCheques", 0) - book.get("usedCheques", 0)
            if remaining >= 3:  # Au moins 3 chèques pour les tests
                active_book = book
                break

    if active_book:
        print_header("Tests d'écriture")
        print_info(f"Utilisation du carnet: {active_book.get('companyName')} / {active_book.get('bankName')}")

        # Demander confirmation
        response = input("\n⚠️  Ces tests vont consommer des numéros de chèques. Continuer? (o/N): ")

        if response.lower() == 'o':
            test_print_cheque(active_book.get("companyName"), active_book.get("bankName"))
            test_print_batch(active_book.get("companyName"), active_book.get("bankName"))
        else:
            print_info("Tests d'écriture ignorés")
    else:
        print_header("Tests d'écriture")
        print_error("Aucun carnet actif avec assez de chèques trouvé")
        print_info("Créez un carnet de chèques dans l'app CheqPrint")

    print_header("Tests terminés")

if __name__ == "__main__":
    main()
